### PR TITLE
Sync ABMini

### PR DESCRIPTION
**Pull request by smialko**
**Overview**:  I adjusted and made changes to enable the app to sync data from ABMini. The main challenge was a different data structure than the one that we got from AB3. 
It wasn't tested on AB3 (it'll be tested on 27th of march), but the code is ready to be reviewed. 

More info: 
https://trello.com/c/9YCYthgV/879-sync-abm

### CODE REVIEW

**Review by jankrzempek — CHANGES_REQUESTED** [review]
Few things to consider and print rule violations. BUT, it's working well I think, really great job! 🥳

File: AirCasting/SDSync/Parsing/MiniSDCardMeasurementsParser.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                }
+                guard let lastKnownUUID else { return }
+                
+                guard let measurements = parseMeasurement(lineString: lineString,
+                                                          sessionUUID: lastKnownUUID) else { return }
+                action(measurements)
+            case .endOfFile: break
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/SDSync/Parsing/MiniSDCardMeasurementsParser.swift:31]: I went to far in here, but maybe we can merge those two last guard's?

File: AirCasting/SDSync/Parsing/MiniSDCardMeasurementsParser.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    
+    func getMeasurementTime(lineString: String) -> Date? {
+        let measurementInfo = lineString.split(separator: ",")
+        guard measurementInfo.count == numberOfColumnsInTheFile else {
+            Log.warning("Line corrupted: \(lineString)")
+            return nil
+        }
+        return SDParsingUtils.dateFrom(date: measurementInfo[MiniSDCardCSVFileFactory.Header.date],
+                                       time: measurementInfo[MiniSDCardCSVFileFactory.Header.time])
+    }
```
[the remark this answers is not in the document]
Comment (smialko) [reply AirCasting/SDSync/Parsing/MiniSDCardMeasurementsParser.swift:75]: Well, I did see the repetition and was thinking about extracting it, yet I don't see it as more readable in this case tbh. For now I'm staying with what we've got

File: AirCasting/Utils/Bluetooth/BluetoothManager.swift [hunk 3]
```diff
@@ -374,6 +374,12 @@ final class BluetoothManager: NSObject, BluetoothCommunicator, CBCentralManagerD
         characteristicsDicoveryCallbacks.removeValue(forKey: peripheral)
     }
     
+    func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
+        if error != nil {
+            print("[SD Sync] \(error)")
```
[the remark this answers is not in the document]
Comment (smialko) [reply AirCasting/Utils/Bluetooth/BluetoothManager.swift:379]: Lols, I changed it to Log.error but it worked fine for me

File: AirCasting/SDSync/MobileSessions/SDCardMobileSessionsSavingService.swift [hunk 4]
```diff
@@ -112,52 +111,44 @@ class SDCardMobileSessionsSavingService: SDCardMobileSessionssSaver {
...
-                            return
-                        }
+                
+                context.perform {
+                    // This happens only with the first line
+                    if sessionData.averaging == nil {
+                        sessionData.averaging = self.averagingService.calculateAveragingWindow(startTime: sessionData.startTime ?? measurements.date, lastMeasurement: lastMeasurementTime)
+                    }
+                    
+                    Log.info("[SD sync] \(i) - \(savedLines): \(measurements.date)") // TO DELETE AFTER TESTING
```
Comment (jankrzempek) [inline AirCasting/SDSync/MobileSessions/SDCardMobileSessionsSavingService.swift:131]: Please bear in mind before merging to develop \ or after testing with Michael - when omit 😄

**Review by jankrzempek — APPROVED** [review]
Great job 🥳

File: AirCasting/SDSync/MobileSessions/SDCardMobileSessionsSavingService.swift (hunk 4 above)
Comment (smialko) [reply AirCasting/SDSync/MobileSessions/SDCardMobileSessionsSavingService.swift:131]: oh thanks, removing it now

**Pull request merged**